File: services/ws-modules/java-data1/src/main/java/au/edu/curtin/et/Data1.java
package au.edu.curtin.et;

import org.teavm.jso.JSBody;
import org.teavm.jso.JSExport;
import org.teavm.jso.JSObject;
import org.teavm.jso.core.JSPromise;
import org.teavm.jso.function.JSConsumer;

public final class Data1 {

    @JSBody(params = {"msg"}, script = "host.log(msg);")
    static native void log(String msg);

    @JSBody(params = {"msg"}, script = "host.setStatus(msg);")
    static native void setStatus(String msg);

    @JSBody(script = "return host.getWsUrl();")
    static native String getWsUrl();

    @JSBody(params = {"url"}, script = "host.wsConnect(url);")
    static native void wsConnect(String url);

    @JSBody(script = "host.wsDisconnect();")
    static native void wsDisconnect();

    @JSBody(script = "return host.wsGetState();")
    static native String wsGetState();

    @JSBody(script = "return host.wsGetAgentId();")
    static native String wsGetAgentId();

    @JSBody(params = {"ms"}, script = "return host.sleep(ms);")
    static native JSPromise<JSObject> sleep(int ms);

    @JSBody(params = {"url", "body"}, script = "return host.putFile(url, body);")
    static native JSPromise<JSObject> putFile(String url, String body);

    @JSBody(params = {"url"}, script = "return host.getFile(url);")
    static native JSPromise<JSObject> getFile(String url);

    @JSBody(script = "return new Date().toISOString();")
    static native String getIsoTimestamp();

    @JSBody(params = {"msg"}, script = "return new Error(msg);")
    static native JSObject jsError(String msg);

    @JSBody(params = {"obj"}, script = "return String(obj);")
    static native String jsObjectToString(JSObject obj);

    @JSExport
    public static JSPromise<JSObject> run() {
        return new JSPromise<>((resolve, reject) -> runAsync(resolve, reject));
    }

    private static void runAsync(JSConsumer<JSObject> resolve, JSConsumer<Object> reject) {
        log("[java-data1] entered run()");
        setStatus("[java-data1] entered run()");
        wsConnect(getWsUrl());
        waitForConnected(0, resolve, reject);
    }

    private static void waitForConnected(int attempt, JSConsumer<JSObject> resolve, JSConsumer<Object> reject) {
        if (attempt >= 100) {
            reject.accept(jsError("Timeout waiting for WebSocket connection"));
            return;
        }
        if ("connected".equals(wsGetState())) {
            waitForAgentId(0, resolve, reject);
            return;
        }
        sleep(100).then(v -> {
            waitForConnected(attempt + 1, resolve, reject);
            return null;
        });
    }

    private static void waitForAgentId(int attempt, JSConsumer<JSObject> resolve, JSConsumer<Object> reject) {
        if (attempt >= 100) {
            reject.accept(jsError("Timeout waiting for agent_id"));
            return;
        }
        String agentId = wsGetAgentId();
        if (agentId != null && !agentId.isEmpty()) {
            String msg = "[java-data1] connected as " + agentId;
            log(msg);
            setStatus(msg);
            doStore(agentId, resolve, reject);
            return;
        }
        sleep(100).then(v -> {
            waitForAgentId(attempt + 1, resolve, reject);
            return null;
        });
    }

    private static void doStore(String agentId, JSConsumer<JSObject> resolve, JSConsumer<Object> reject) {
        String filename = "test_data.txt";
        String testContent = "Hello from java-data1 at " + getIsoTimestamp() + "!";
        String storageUrl = "/storage/" + agentId + "/" + filename;
        String msg = "[java-data1] storing data to " + storageUrl;
        log(msg);
        setStatus(msg);
        putFile(storageUrl, testContent).then(v -> {
            doFetch(storageUrl, testContent, resolve, reject);
            return null;
        });
    }

    private static void doFetch(
            String storageUrl, String testContent, JSConsumer<JSObject> resolve, JSConsumer<Object> reject) {
        String msg = "[java-data1] fetching data from " + storageUrl;
        log(msg);
        setStatus(msg);
        getFile(storageUrl).then(result -> {
            verifyAndFinish(testContent, jsObjectToString(result), resolve, reject);
            return null;
        });
    }

    private static void verifyAndFinish(
            String testContent, String retrieved, JSConsumer<JSObject> resolve, JSConsumer<Object> reject) {
        if (testContent.equals(retrieved)) {
            String ok = "[java-data1] VERIFICATION SUCCESS - data matches!";
            log(ok);
            setStatus(ok);
        } else {
            String fail = "[java-data1] VERIFICATION FAILURE\nSent: " + testContent + "\nGot: " + retrieved;
            log(fail);
            setStatus(fail);
            reject.accept(jsError("Data mismatch"));
            return;
        }
        sleep(2000).then(v -> {
            wsDisconnect();
            String done = "[java-data1] workflow complete";
            log(done);
            setStatus(done);
            resolve.accept(null);
            return null;
        });
    }
}

File: services/ws-modules/kotlin-data1/src/wasmJsMain/kotlin/Data1.kt
// Kotlin 2.x still stability-gates the whole wasmJs interop surface (js(), JsAny, Promise) behind this
// opt-in; every declaration in this file is an interop bridge, so the opt-in is file-scoped.
@file:OptIn(ExperimentalWasmJsInterop::class)

package au.edu.curtin.et

import kotlin.js.ExperimentalWasmJsInterop
import kotlin.js.Promise

// The pkg/ shim installs a `host` global carrying the browser-side WebSocket/storage/console operations, the
// same contract java-data1's TeaVM @JSBody bindings use. Each js() body below is a single-expression bridge.
private fun hostLog(msg: String): Unit = js("host.log(msg)")

private fun hostSetStatus(msg: String): Unit = js("host.setStatus(msg)")

private fun hostGetWsUrl(): String = js("host.getWsUrl()")

private fun hostWsConnect(url: String): Unit = js("host.wsConnect(url)")

private fun hostWsDisconnect(): Unit = js("host.wsDisconnect()")

private fun hostWsGetState(): String = js("host.wsGetState()")

private fun hostWsGetAgentId(): String = js("host.wsGetAgentId()")

private fun hostSleep(ms: Int): Promise<JsAny?> = js("host.sleep(ms)")

private fun hostPutFile(url: String, body: String): Promise<JsAny?> = js("host.putFile(url, body)")

private fun hostGetFile(url: String): Promise<JsAny?> = js("host.getFile(url)")

private fun isoTimestamp(): String = js("new Date().toISOString()")

private fun jsToString(value: JsAny?): String = js("String(value)")

private fun resolvedPromise(): Promise<JsAny?> = js("Promise.resolve(null)")

private fun rejectedPromise(msg: String): Promise<JsAny?> = js("Promise.reject(new Error(msg))")

private fun installRun(f: () -> Promise<JsAny?>): Unit = js("globalThis.kotlinData1Run = f")

private fun status(msg: String) {
    hostLog("[kotlin-data1] $msg")
    hostSetStatus("[kotlin-data1] $msg")
}

private fun waitUntil(what: String, attempt: Int = 0, ready: () -> Boolean): Promise<JsAny?> = when {
    ready() -> resolvedPromise()
    attempt >= 100 -> rejectedPromise("Timeout waiting for $what")
    else -> hostSleep(100).then<JsAny?> { waitUntil(what, attempt + 1, ready) }
}

private fun storeAndVerify(agentId: String): Promise<JsAny?> {
    val content = "Hello from kotlin-data1 at ${isoTimestamp()}!"
    val storageUrl = "/storage/$agentId/test_data.txt"
    status("storing data to $storageUrl")
    return hostPutFile(storageUrl, content)
        .then<JsAny?> {
            status("fetching data from $storageUrl")
            hostGetFile(storageUrl)
        }
        .then<JsAny?> { retrieved ->
            val got = jsToString(retrieved)
            if (got == content) {
                status("VERIFICATION SUCCESS - data matches!")
                null
            } else {
                status("VERIFICATION FAILURE - data mismatch!\nSent: $content\nGot: $got")
                rejectedPromise("Data mismatch")
            }
        }
}

private fun runWorkflow(): Promise<JsAny?> {
    status("entered run()")
    hostWsConnect(hostGetWsUrl())
    return waitUntil("WebSocket connection") { hostWsGetState() == "connected" }
        .then<JsAny?> { waitUntil("agent_id") { hostWsGetAgentId().isNotEmpty() } }
        .then<JsAny?> {
            val agentId = hostWsGetAgentId()
            status("connected as $agentId")
            storeAndVerify(agentId)
        }
        .then<JsAny?> { hostSleep(2000) }
        .then<JsAny?> {
            hostWsDisconnect()
            status("workflow complete")
            null
        }
}

fun main() {
    installRun { runWorkflow() }
}
